### PR TITLE
add custom http headers

### PR DESCRIPTION
**Pull request by GuyAfik**
## Status
Ready

## Related Issues
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-4509)

## Description
add the ability to add custom http headers that will be sent in each http request
implements https://github.com/demisto/demisto-py/issues/103

## Must have
- [x] Unit Test or Example Code
- [x] Changelog entry

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,9 @@
 [PyPI History][1]
 [1]: https://pypi.org/project/demisto-py/#history
 
+#3.2.9
+* Added the ability to use custom headers for each request.
```
Comment (dorschw) [inline CHANGELOG.md:7]:
```suggestion
* Added the ability to add custom request headers.
```

**Comment by GuyAfik** [comment]
@dorschw @JudahSchwartz do we need to validate the form of the environment variable to be `header1=value1,header2=header2`?

File: demisto_client/__init__.py [hunk 2]
```diff
@@ -61,6 +63,11 @@ def configure(base_url=None, api_key=None, verify_ssl=None, proxy=None, username
         username = os.getenv('DEMISTO_USERNAME')
     if password is None:
         password = os.getenv('DEMISTO_PASSWORD')
+    if additional_headers is None:
+        if headers := os.getenv('DEMISTO_HTTP_HEADERS'):
+            additional_headers = dict(header.split('=') for header in headers.split(','))
```
Comment (dorschw) [inline demisto_client/__init__.py:68]: let's delight the customers by adding some try-except block when an invalid format is used

File: tests/mocks_test.py [hunk 3]
```diff
@@ -20,6 +20,22 @@ def assert_reset():
     assert len(responses.calls) == 0
 
 
+def get_login_mocker(mocker, num_of_http_requests_mock=0):
```
Comment (dorschw) [inline tests/mocks_test.py:23]:
```suggestion
def get_login_mocker(mocker, num_of_http_requests_mock: int=0):
```

**Review by dorschw — COMMENTED** [review]
Very nice

File: demisto_client/__init__.py (hunk 2 above)
Comment (GuyAfik) [reply demisto_client/__init__.py]: done

File: demisto_client/__init__.py [hunk 4]
```diff
@@ -17,8 +19,11 @@
     __version__ = 'dev'
 
 
+DEMISTO_HTTP_HEADERS_REGEX_PATTERN = r'^\w+=\w+(,\w+=\w+)*$'
```
Comment (JudahSchwartz) [inline demisto_client/__init__.py:22]: are we sure we want to limit to \w? The customer's usecase is for authentication, theres no special chars needed?

File: demisto_client/__init__.py [hunk 5]
```diff
@@ -61,6 +68,16 @@ def configure(base_url=None, api_key=None, verify_ssl=None, proxy=None, username
         username = os.getenv('DEMISTO_USERNAME')
     if password is None:
         password = os.getenv('DEMISTO_PASSWORD')
+    if additional_headers is None:
+        if headers := os.getenv('DEMISTO_HTTP_HEADERS'):
+            if re.match(DEMISTO_HTTP_HEADERS_REGEX_PATTERN, headers):
+                additional_headers = dict(header.split('=') for header in headers.split(','))
```
Comment (JudahSchwartz) [inline demisto_client/__init__.py:74]: lets strip the whitespace around keys and vals

**Review by JudahSchwartz — APPROVED** [review]
nice! 2 minor comments

File: demisto_client/__init__.py (hunk 4 above)
Comment (GuyAfik) [reply demisto_client/__init__.py:22]: you are right, ill update regex to include special chars as well.

Comment (JudahSchwartz) [reply demisto_client/__init__.py:22]: we should also support the case where it is ""

Comment (GuyAfik) [reply demisto_client/__init__.py:22]: do you mean empty string?

Comment (JudahSchwartz) [reply demisto_client/__init__.py:22]: or even " " should ideally not break it

Comment (GuyAfik) [reply demisto_client/__init__.py:22]: why do we need to support this in regex?

**Comment by GuyAfik** [comment]
@JudahSchwartz please re-review as regex was updated.

File: demisto_client/__init__.py (hunk 4 above)
Comment (JudahSchwartz) [reply demisto_client/__init__.py]: if someone has a script that looks like 

```
#setting my envVariables
export Demisto_x..= hello
export DEMISTO_HEADERS= 
export ANOTHWR_HEADING=

```
it shouldnt break

Comment (GuyAfik) [reply demisto_client/__init__.py]: i didn't add it to the regex cause it does make sense, added it as a separated check to handle empty strings as well.

**Review by JudahSchwartz — APPROVED** [review]
nice

**Pull request merged**